Reflectometry: Multiple Row Checks

DavidFair opened the pull request:
**Description of work.**
    Handle multiple row matches in exp. settings
    
    This adds handling for if we match multiple theta AND title values.
    Since we don't want to write our own checking logic in the view, we do
    this once a user hits process so that we can check the regex values.
    
    This is typically seen if a user has a greedy regex (e.g. ".*") and a
    more specific matcher than conflict.

**To test:**
- Enter a group name `A`, enter an angle of `0.1` and the run as `13460`
- Copy paste the group, rename the other groups to `AA` and `AAA`
- In experiment settings tab add another row. In the first set the title to `A.*` and the angle to 0.1, in the second set the title to `AA.*` and angle 0.1
- Try processing in the main tab this should give a sensible warning
- In the experiment settings tab, try changing the title matchers to `AA.*` and `AAA.*`
- This should process the first run (using the defaults) and give a warning on the second

Fixes #33516 

*This does not require release notes* because this is part of ongoing, unreleased work in #31795 

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Handle multiple row matches in exp. settings

This adds handling for if we match multiple theta AND title values.
Since we don't want to write our own checking logic in the view, we do
this once a user hits process so that we can check the regex values.

This is typically seen if a user has a greedy regex (e.g. ".*") and a
more specific matcher than conflict.

Re #33516

DavidFair committed:
Replace new with make_unique

This is the correct way to do it, and also indicates we will manage the
lifetime too.

DavidFair committed:
Move error handling for multiple regex matches to job manager

Also set the error on the row rather than a pop-up message. This will
stop autoreduction from halting.

Make sure row state is updated after catching error

Add test to BatchJobManagerProcessingTest

Injects a mocking point by using IBatch in a forwarding class, similar to the preview rows.
This allows us to simply mock rather than wire up experiment for a given
row calling createConfiguredAlgorithm

Adds a new test to check that a MultipleRowsException is handled and
marks the row in an error state correctly.

Re #33516

gemmaguest reviewed (changes requested):
Code looks good :+1:

I've found a couple of issues with testing:
- There's a minor problem in the first test example where 2 rows are failing and the state is not updated on the GUI for the second one. This is because `BatchPresenter::notifyBatchComplete` doesn't call `notifyReductionResumed` before it calls `startBatch` (previously there would have been no opportunity for row states to change between executing a previous batch and running the next one, but with the new error handling they can now change). Technically we only need to call `m_runsPresenter->notifyRowStateChanged` but maybe calling `notifyReductionResumed` follows the existing pattern better. 
- There's an unhandled exception when running autoreduction when multiple regexes match. This needs further investigation.

gemmaguest commented:
I was a little confused by the test instructions but these are the results I'm getting, which look correct to me. 

**Test 1**
- Group `A` matches `A.*` and picks up the dq/q setting of `0.01`
- The other groups match both regexes so fail

![image](https://user-images.githubusercontent.com/12895056/155713576-06fa9a80-7c95-4081-85b9-c0e93da8f18d.png)
![image](https://user-images.githubusercontent.com/12895056/155713588-2b1bb301-99fd-427f-8c95-ae3bbd464e8c.png)

**Test 2**
- Group `A` does not match either regex so has its default dq/q of `0.238`
- Group `AA` matches `AA.*` so picks up dq/q of `0.01`
- Group `AAA` matches both regexes so fails.

![image](https://user-images.githubusercontent.com/12895056/155713731-c3ee816e-0220-4c2b-87e8-5b752f2cc992.png)
![image](https://user-images.githubusercontent.com/12895056/155713740-835a9ca6-c770-43d7-9785-7ed06b00fbfc.png)

gemmaguest committed:
Fix a bug where autoreduction would end early

If a group only contains rows that fail due to the new error that is
caught, then an empty algorithms list would be returned. This would
cause autoreduction to assume that everything had been processed even
though there could be remaining groups to process. This commit ensures
we skip over groups that do not contain any valid rows.

Re #33516

gemmaguest committed:
Fix a bug where the GUI was not updated with the row state

Re #33516

gemmaguest committed:
Rename new functions to be more consistent with existing

Re #33516

gemmaguest reviewed (approved):
David is away so I've made the suggested changes myself. The autoreduction bug was caused by returning an empty algorithms list for groups which do not contain any valid rows, so I've added a check to skip those groups. I've also added a fix to refresh the GUI before continuing to the next batch to ensure any errors are displayed, and renamed the two new functions for consistency.

This all now works as expected. This PR adds some important error checking to ensure that reduction is not performed if it is ambiguous which settings would be used.
